### PR TITLE
DF 1704 - ReOpen PoWE and set correct nextSD, when it is an active CP…

### PR DESCRIPTION
**Pull request by emina-str**
… (on updating PoWE)

https://alv-ch.atlassian.net/browse/DF-1704

### CODE REVIEW

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java [hunk 1]
```diff
@@ -375,21 +386,23 @@ private ChangeLog recalculateDates(DateCalculator dateCalculator) {
 		ChangeLog changeLog = new ChangeLog();
 		LocalDate startDate = dateCalculator.calculateStartDate(this.ownerUserId, this.getControlPeriod());
 		changeLog.checkChanges("start-date", this.startDate, startDate, newValue -> this.startDate = newValue);
-		if (this.status.isInAnyStates(DRAFT, OPEN)) {
-			ProofOfWorkEffortsStatus newStatus = this.calculateInitialStatus(startDate);
+
+		if (this.status.isInAnyStates(DRAFT, OPEN) || (this.status == SUBMITTED && !this.controlPeriod.isInThePast())) {
```
Comment (pavax) [inline domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java]: suggestion -> create two checks

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java [hunk 2]
```diff
@@ -375,21 +386,23 @@ private ChangeLog recalculateDates(DateCalculator dateCalculator) {
...
-	Optional<WorkEffort> findWorkEffort(WorkEffortId workEffortId) {
-		return this.getWorkEfforts()
-				.stream()
-				.filter(workEffort -> workEffort.getId().equals(workEffortId))
-				.findFirst();
+	private ProofOfWorkEffortsStatus resolveStatusOnUpdate(LocalDate startDate) {
+		if (this.status.isInAnyStates(DRAFT, OPEN)) {
+			return this.calculateInitialStatus(startDate);
+		}
+		return RE_OPENED;
```
Comment (pavax) [inline domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java]: I'd use the Status OPEN

Comment (emina-str) [reply domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java]: fixed

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java (hunk 1 above)
Comment (emina-str) [reply domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java]: created.

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java [hunk 3]
```diff
@@ -390,21 +401,23 @@ private ChangeLog recalculateDates(DateCalculator dateCalculator) {
 		ChangeLog changeLog = new ChangeLog();
 		LocalDate startDate = dateCalculator.calculateStartDate(this.ownerUserId, this.getControlPeriod());
 		changeLog.checkChanges("start-date", this.startDate, startDate, newValue -> this.startDate = newValue);
+
+		if (this.status == SUBMITTED && !this.controlPeriod.isInThePast()) {
```
Comment (pavax) [inline domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/ProofOfWorkEfforts.java]: extract a method called hasBeenSubmittedAndIsActive

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/DateCalculator.java [hunk 4]
```diff
@@ -55,10 +55,12 @@ LocalDate calculateSubmissionDate(UserId userId, ControlPeriod controlPeriod) {
 	}
 
 
-	void recalculateNext(ProofOfWorkEfforts changedProofOfWorkEfforts) {
+	void adjustNext(ProofOfWorkEfforts changedProofOfWorkEfforts) {
 		this.proofOfWorkEffortsRepository.findNext(changedProofOfWorkEfforts.getOwnerUserId(), changedProofOfWorkEfforts.getControlPeriod())
 				.filter(next -> isBlockingNext(changedProofOfWorkEfforts, next))
-				.ifPresent((conflicting) -> conflicting.recalculate(this));
+				.ifPresent((conflicting) -> {
```
Comment (pavax) [inline domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/npa/DateCalculator.java]: simplify

**Pull request merged**